R/Methods-FAKinGroupResults.R

```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
+                    expectedAffInKin <- match(expectedAffInKin,
+                                              colnames(kinPhenotyped))
+                    expectedKin <- kinPhenotyped[expectedAffInKin, expectedAffInKin]
+                    diag(expectedKin) <- NA
+                    ## get the table with the frequencies of kinship values...
+                    expectedKinFreqs <- table(as.vector(expectedKin))
+                    observedKinFreqs <- z$kinfreq
+                    if(length(expectedKinFreqs) > 0 & length(observedKinFreqs) > 0){
+                        ## names of the table are the kinship values, ordered increasingly by size.
+                        expKinShipValue <- names(expectedKinFreqs[length(expectedKinFreqs)])
+                        obsKinShipValue <- names(observedKinFreqs[length(observedKinFreqs)])
+                        if(as.numeric(expKinShipValue) >= as.numeric(obsKinShipValue)){
+                            ## if the kinship value is the same, compare the counts.
+                            if(expKinShipValue==obsKinShipValue){
+                                if(expectedKinFreqs[expKinShipValue] >=
+                                   observedKinFreqs[obsKinShipValue])
+                                    Res[2] <- 1
+                            }else{
+                                ## means it's larger, so add +1
                                 Res[2] <- 1
-                        }else{
-                            ## means its larger, so add +1
-                            Res[2] <- 1
+                            }
                         }
                     }
                 }
-            }
-            return(Res)
-        }))
-        return(ExpRes)
-    })
-    ## BigRes is now a list, length nsim, each element being a matrix, ncol=2,
-    ## nrow=length affectedKins.
-    ExpRatio <- colSums(do.call(rbind, lapply(BigRes,
-                                              function(simRes){simRes[, 1]})))
-    ExpKin <- colSums(do.call(rbind, lapply(BigRes,
-                                            function(simRes){simRes[, 2]})))
-    ExpVals <- do.call(rbind, lapply(BigRes, function(simRes){simRes[, 3]}))
-    ExpVals <- split(t(ExpVals), f=names(affectedKins))
-    expDensity <- lapply(ExpVals, density)
-    expHist <- lapply(ExpVals, hist, plot=FALSE)
-    expDensity <- expDensity[names(affectedKins)]
-    expHist <- expHist[names(affectedKins)]
-    names(ExpRatio) <- names(affectedKins)
-    names(ExpKin) <- names(affectedKins)
+                return(Res)
+            }))
+            ExpRatio <- ExpRatio + ExpRes[, 1]
+            ExpKin   <- ExpKin   + ExpRes[, 2]
+            BigRes <- append(BigRes, list(ExpRes[, 3]))
+        }
+        ## BigRes is now a list of length nr.inc.sim, each element a vector
+        ## of length affectedKins.
+        ExpVals <- do.call(rbind, BigRes)
+        ExpVals <- split(t(ExpVals), f=names(affectedKins))
```

Comment:
What exactly do you do here? Is this `split` doing the correct thing? Note that if `ExpVals` is a `matrix` (which can well be with the proposed changes above) `split` will consider it as a `numeric` vector instead. If you want to explicitely split by row (i.e. `f` groups rows in the `matrix`/`data.frame`) you should use `split.data.frame` instead of `split`.

Reply:
This is Jo-code. I just took the same stuff you have done into the loop. It splits properly and as wanted. Doesn't look nice, though. I also wasn't too happy with this weird construct.

Reply:
you mean it still splits properly even after the proposed change above? jo-code is bad - has improved over the last years, but the old one was sometimes not efficient...